x/ccv/provider/keeper/throttle.go

```diff
@@ -0,0 +1,400 @@
+	providertypes "github.com/cosmos/interchain-security/x/ccv/provider/types"
+	ccvtypes "github.com/cosmos/interchain-security/x/ccv/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+// This file contains functionality relevant to the throttling of slash and vsc matured packets, aka circuit breaker logic.
+
+// HandlePendingSlashPackets handles all or some portion of pending slash packets received by consumer chains,
+// depending on throttling logic. The slash meter is decremented appropriately in this method.
+func (k Keeper) HandlePendingSlashPackets(ctx sdktypes.Context) {
+
+	meter := k.GetSlashMeter(ctx)
+
+	// Don't start iterating if meter is 0 or negative in value
+	if !meter.IsPositive() {
+		return
+	}
+
+	handledEntries := []providertypes.SlashPacketEntry{}
+
+	// Iterate through ordered (by received time) slash packet entries from any consumer chain
+	k.IteratePendingSlashPacketEntries(ctx, func(entry providertypes.SlashPacketEntry) (stop bool) {
+
+		// Obtain the validator power relevant to the slash packet that's about to be handled
+		// (this power will be removed via jailing or tombstoning)
+		valPower := k.stakingKeeper.GetLastValidatorPower(ctx, entry.ValAddr)
+
+		// Subtract this power from the slash meter
+		meter = meter.Sub(sdktypes.NewInt(valPower))
+
+		// Handle slash packet by passing in chainID and appropriate callbacks, relevant packet data is deleted in this method
+		k.HandlePacketDataForChain(ctx, entry.ConsumerChainID, k.HandleSlashPacket, k.HandleVSCMaturedPacket)
+
+		// Store handled entry to be deleted after iteration is completed
+		handledEntries = append(handledEntries, entry)
+
+		// Do not handle anymore slash packets if the meter is 0 or negative in value
+		stop = !meter.IsPositive()
+		return stop
+	})
+
+	// Handled entries are deleted after iteration is completed
+	k.DeletePendingSlashPacketEntries(ctx, handledEntries...)
+
+	// Persist current value for slash meter
+	k.SetSlashMeter(ctx, meter)
+}
+
+// HandlePacketDataForChain handles only the first queued slash packet relevant to the passed consumer chainID,
+// and then handles any trailing vsc matured packets in that (consumer chain specific) queue.
+//
+// Note: Any packet data which is handled in this method is also deleted from the (consumer chain specific) queue.
+func (k Keeper) HandlePacketDataForChain(ctx sdktypes.Context, consumerChainID string,
+	slashPacketHandler func(sdktypes.Context, string, ccvtypes.SlashPacketData) (bool, error),
+	vscMaturedPacketHandler func(sdktypes.Context, string, ccvtypes.VSCMaturedPacketData),
+) {
+
+	// Store ibc sequence numbers to delete data after iteration is completed
+	seqNums := []uint64{}
+
+	idx := 0
+	k.IteratePendingPacketData(ctx, consumerChainID, func(ibcSeqNum uint64, data interface{}) (stop bool) {
+
+		switch data := data.(type) {
+
+		case ccvtypes.SlashPacketData:
+			if idx == 0 {
+				_, err := slashPacketHandler(ctx, consumerChainID, data)
```

Comment:
Check that the slash packet is the expected one, for which the corresponding packet was already added to the slash meter.

Reply:
You are right. Maybe it's not a good idea to pollute the code with a lot of panic messages. Maybe it would be cool to have some kind of assert that is enabled only on testnests. Does the SDK provide this functionality?

Reply:
This would be very useful, but I'm not aware of such functionality besides logging